File: Niveau 4/3 – Structures de données et Balayages/06 - Peinture.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

int main()
{
    ios::sync_with_stdio(false);

    int nbQuantites, quantiteRecherchee;
    cin >> nbQuantites >> quantiteRecherchee;

    vector<int> quantites;

    for (int i = 0; i < nbQuantites; i++)
    {
        int quantite;
        cin >> quantite;
        quantites.push_back(quantite);

        if (quantite * 2 == quantiteRecherchee)
        {
            cout << "OUI";
            return 0;
        }
        if (i > 0 && quantite + quantites[0] > quantiteRecherchee)
        {
            break;
        }
    }

    for (int element : quantites)
    {
        int toFind = quantiteRecherchee - element;
        if (find(quantites.begin(), quantites.end(), toFind) != quantites.end())
        {
            cout << "OUI";
            return 0;
        }
    }

    cout << "NON";
    return 0;
}

File: Niveau 4/3 – Structures de données et Balayages/06 - Peinture.py
def find(n, arr):
  start = 0
  end = len(arr) - 1
  m = 0
  while start <= end:
    m = (start + end) // 2
    if(arr[m] == n):
      return True
    else:
      if(n > arr[m]):
        start = m + 1
      else:
        end = m - 1
  
  return False

def main():
  nbQuantites, quantiteRecherchee = map(int, input().split())

  quantites = list(map(int, input().split()))

  f = False

  for quantite in quantites:
    if(quantite * 2 == quantiteRecherchee):
      f = True
      break
    toFind = quantiteRecherchee - quantite
    if(find(toFind, quantites)):
      f = True
      break
    if(quantite + quantites[0] > quantiteRecherchee):
      break
  
  print("OUI" if f else "NON")

main()